### PR TITLE
Implement in-memory coordination backend reference spec

### PR DESCRIPTION
**Pull request by ahrav**
## Summary

Implement the in-memory coordination backend — the executable specification
for the shard coordination protocol. This is the reference implementation
that all production backends (FoundationDB, Postgres, etc.) must match
in observable behavior.

**Impact**: 6 files changed (+2,081 lines), 23 in_memory tests + 4 property tests, 253 total coordination tests passing.

### Source Changes

- **`coordination/traits.rs`** (new, 363 lines): The `CoordinationBackend`
  trait — the synchronous contract for all backends. Defines 7 operations:
  `acquire_and_restore`, `renew`, `checkpoint`, `complete`, `park_shard`,
  `split_replace`, `split_residual`. Extensive rustdoc documenting behavior,
  idempotency semantics, and invariants with academic references (Kleppmann
  2016, Gray & Cheriton 1989, FoundationDB SIGMOD 2021).

- **`coordination/in_memory.rs`** (new, 1,661 lines): `InMemoryCoordinator` —
  single-threaded HashMap-based implementation enforcing all 10 ShardRecord
  invariants. Highlights:
  - Fencing tokens (Kleppmann 2016) via monotonic `fence_epoch`
  - Time-bounded leases (Gray & Cheriton 1989)
  - Bounded idempotency (Stripe pattern) with 16-entry FIFO op-log
  - Remove-mutate-restore pattern for split operations (resolves Rust
    borrow checker conflict with simultaneous parent mutation and child
    insertion)
  - Defense-in-depth replay detection for `split_residual` via
    spawned-probe secondary check after op-log eviction
  - Tiger-style `assert_invariants()` on every mutation path

- **`coordination/shard_spec.rs`** (modified): Added `ParentCursorOutOfBounds`
  variant to `SplitValidationError` — ensures a residual split doesn't strand
  the parent's cursor outside its new key range.

- **`coordination/mod.rs`** (modified): Re-exports for `InMemoryCoordinator`,
  `CoordinationBackend`, `SplitReplaceResult`, `SplitResidualResult`.

### Test Coverage

| Test Type | Count | What It Covers |
|-----------|-------|----------------|
| Unit tests | 19 | All 7 protocol operations, error paths, lease expiry, fence stale rejection, op-log eviction, split replay, cursor bounds |
| Property tests | 4 | Fence monotonicity, cursor monotonicity, idempotent replay across all 3 idempotent ops, random operation sequence invariant preservation |

## Why These Changes

The coordination backend is the core of the distributed shard assignment
protocol. The in-memory reference implementation serves three purposes:

1. **Executable specification**: Codifies the protocol rules so production
   backends have an unambiguous behavioral target.
2. **Deterministic simulator foundation**: The synchronous, time-as-input
   design (D2.13, D2.17) enables the future deterministic simulator to
   exercise the protocol without an async runtime.
3. **Property-based verification**: The `random_ops_preserve_invariants`
   test exercises arbitrary sequences of operations and asserts all 10
   ShardRecord invariants hold after every step.

## Architecture

```
┌──────────────────────────────────────────────┐
│ CoordinationBackend (trait)                   │
│   acquire_and_restore                        │
│   renew / checkpoint / complete / park_shard  │
│   split_replace / split_residual             │
└───────────────┬──────────────────────────────┘
                │ impl
                ▼
┌──────────────────────────────────────────────┐
│ InMemoryCoordinator                          │
│   HashMap<(TenantId, ShardKey), ShardRecord> │
│   Fencing tokens · Leases · Op-log           │
│   Tiger-style assert_invariants()            │
└──────────────────────────────────────────────┘
                │ future impls
                ▼
        FoundationDB / Postgres / Simulator
```

### Shard State Machine

```
       ┌──────────────┐
       │    Active     │
       └──┬───┬────┬──┘
          │   │    │
 complete │   │    │ park_shard
   ┌──────┘   │    └───────┐
   ▼    split_replace      ▼
┌──────┐      │       ┌────────┐
│ Done │      ▼       │ Parked │
└──────┘  ┌───────┐   └────────┘
          │ Split  │
          └───────┘
```

## Design Decisions

| ID | Decision | Rationale |
|----|----------|-----------|
| D2.13 | Synchronous trait (no async) | Keeps deterministic simulator simple; async is the backend's concern |
| D2.14 | Lease-gated ops take `(TenantId, Lease)` | Fencing token protocol — all writes carry fence epoch |
| D2.15 | `acquire_and_restore` is the only lease-free op | Entry point for workers; all others require valid lease |
| D2.16 | Operation-specific error enums via `From<CoordError>` | Type-safe error handling per operation |
| D2.17 | `now: LogicalTime` passed explicitly | Coordinator never reads a clock — essential for DST |

## How Has This Been Tested?

```
cargo test --lib -p gossip-contracts coordination --
# 253 passed; 0 failed; 0 ignored

cargo clippy -p gossip-contracts --all-targets --all-features -- -D warnings
# Clean
```

### Key Test Scenarios

- **acquire**: basic, not-found, already-leased, after-expiry, terminal-rejected
- **renew**: basic, stale-fence-rejected
- **checkpoint**: basic, op-id-conflict
- **complete**: basic (terminal transition)
- **park**: basic (terminal transition)
- **split_replace**: basic, child-id-determinism, idempotent-replay
- **split_residual**: basic, cursor-out-of-bounds rejection, replay-via-spawned-after-eviction
- **cross-cutting**: only-latest-fence-holder-can-mutate, op-log-eviction-treats-old-op-as-new
- **property-based**: fence monotonicity (strict >), cursor monotonicity, idempotent replay across all 3 ops with randomized payloads, random operation sequences preserve all invariants

## Breaking Changes

None. All changes are additive — new modules and re-exports. No existing
public API signatures changed.

## Reviewer Focus Areas

1. **Remove-mutate-restore pattern** in `split_replace`/`split_residual`
   (`in_memory.rs:364-399`) — borrow checker workaround for simultaneous
   parent mutation and child insertion
2. **Spawned-probe replay detection** for `split_residual` after op-log
   eviction — the secondary defense when the FIFO evicts the original entry
3. **Property test: `random_ops_preserve_invariants`** — strongest correctness
   evidence, exercising arbitrary operation sequences

## Test plan

- [ ] `cargo test --lib -p gossip-contracts coordination` — all 253 tests pass
- [ ] `cargo clippy -p gossip-contracts --all-targets --all-features -- -D warnings` — clean
- [ ] Review property tests for adequate invariant coverage
- [ ] Verify split operations handle edge cases (eviction, replay, cursor bounds)

### CODE REVIEW

File: crates/gossip-contracts/src/coordination/in_memory.rs [hunk 1]
```diff
@@ -0,0 +1,1661 @@
...
+    fn arb_op() -> impl Strategy<Value = Op> {
+        prop_oneof![
+            (0u8..4).prop_map(|w| Op::Acquire { worker: w }),
+            (b'a'..b'y').prop_map(|k| Op::Checkpoint { cursor_key: k }),
+            (b'a'..b'y').prop_map(|k| Op::Complete { cursor_key: k }),
+            Just(Op::Park),
+            Just(Op::Renew),
+            (1u64..200).prop_map(|t| Op::TimeAdvance { ticks: t }),
+        ]
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-contracts/src/coordination/in_memory.rs]: Fixed — added `SplitReplace` and `SplitResidual` variants to the `Op` enum in `random_ops_preserve_invariants`.

Both use a fixed split point (`[a,m) + [m,z)`) against the test spec `[a,z)`. Most split attempts will fail (wrong status, cursor out of bounds after prior ops, etc.), but the property test only cares that invariants hold after every operation — which is exactly the point. The weighted selection gives splits lower frequency than checkpoints/acquires since they're less common in practice.

All 387 tests pass including the extended property test.
